src/aes_hash.rs

```diff
@@ -22,6 +22,74 @@ pub struct AHasher {
     key: u128,
 }
 
+#[cfg(any(
+    all(any(target_arch = "x86", target_arch = "x86_64"), target_feature = "aes", not(miri)),
+    all(target_arch = "aarch64", target_feature = "aes", not(miri)),
+    all(feature = "nightly-arm-aes", target_arch = "arm", target_feature = "aes", not(miri)),
+))]
+fn hash_batch_128b(data: &mut &[u8], hasher: &mut AHasher) {
+    let tail = data.read_last_u128x8();
+    let current = [
+        aesenc(hasher.key, tail[0]),
+        aesdec(hasher.key, tail[1]),
+        aesenc(hasher.key, tail[2]),
+        aesdec(hasher.key, tail[3]),
+        aesenc(hasher.key, tail[4]),
+        aesdec(hasher.key, tail[5]),
+        aesenc(hasher.key, tail[6]),
+        aesdec(hasher.key, tail[7]),
+    ];
+    let mut current = [
+        convert_u128_to_vec256(current[0], current[1]),
+        convert_u128_to_vec256(current[2], current[3]),
+        convert_u128_to_vec256(current[4], current[5]),
+        convert_u128_to_vec256(current[6], current[7]),
+    ];
+    let mut sum: [Vector256; 2] = [
+        convert_u128_to_vec256(hasher.key, !hasher.key), 
+        convert_u128_to_vec256(hasher.key, !hasher.key)
+    ];
+    let tail = [
+        convert_u128_to_vec256(tail[0], tail[1]),
+        convert_u128_to_vec256(tail[2], tail[3]),
+        convert_u128_to_vec256(tail[4], tail[5]),
+        convert_u128_to_vec256(tail[6], tail[7]),
+    ];
+    sum[0] = add_by_64s_vec256(sum[0], tail[0]);
+    sum[0] = add_by_64s_vec256(sum[0], tail[1]);
+    sum[1] = shuffle_and_add_vec256(sum[1], tail[2]);
+    sum[1] = shuffle_and_add_vec256(sum[1], tail[3]);
+    while data.len() > 128 {
```

Comment:
We could go even further to preserve the algorithm. In existing code it is doing 4 blocks of 128 bits in parallel, Here it is doing 4 blocks each of which is 2 inner blocks of 128 which are parallelized by the vaes instruction. But could instead make it 2 blocks each of which is 2 inner blocks parallelized by vaes which would make the output identical with and without the vaes instruction. This would make it much easier to audit. It would also eliminate the need for the additional `if` check which will improve the speed of variable length strings.

Reply:
I tested before with `aesenc`, the AMD Zen3 CPU at least performs worse with `vaes` when processing same amount data with classical `aes`. The performance benefit does not appear unless with larger batch sizes.

Reply:
I do see some speedup with
```rust
 if data.len() > 64 {
                    let tail = data.read_last_u128x4();
                    let mut current: [u128; 4] = [self.key; 4];
                    current[0] = aesenc(current[0], tail[0]);
                    current[1] = aesdec(current[1], tail[1]);
                    current[2] = aesenc(current[2], tail[2]);
                    current[3] = aesdec(current[3], tail[3]);
                    let mut sum: [u128; 2] = [self.key, !self.key];
                    sum[0] = add_by_64s(sum[0].convert(), tail[0].convert()).convert();
                    sum[1] = add_by_64s(sum[1].convert(), tail[1].convert()).convert();
                    sum[0] = shuffle_and_add(sum[0], tail[2]);
                    sum[1] = shuffle_and_add(sum[1], tail[3]);
                    let mut sum = convert_u128_to_vec256(sum[0], sum[1]);
                    let mut current = [
                        convert_u128_to_vec256(current[0], current[1]),
                        convert_u128_to_vec256(current[2], current[3]),
                    ];
                    while data.len() > 64 {
                        let (blocks, rest) = read2_vec256(data);
                        current[0] = aesdec_vec256(current[0], blocks[0]);
                        current[1] = aesdec_vec256(current[1], blocks[1]);
                        sum = shuffle_and_add_vec256(sum, blocks[0]);
                        sum = shuffle_and_add_vec256(sum, blocks[1]);
                        data = rest;
                    }
                    let current = [convert_vec256_to_u128(current[0]), convert_vec256_to_u128(current[1])];
                    let sum = convert_vec256_to_u128(sum);
                    self.hash_in_2(current[0][0], current[0][1]);
                    self.hash_in_2(current[1][0], current[1][1]);
                    self.hash_in_2(sum[0], sum[1]);
                } else {
                    //len 33-64
                    let (head, _) = data.read_u128x2();
                    let tail = data.read_last_u128x2();
                    self.hash_in_2(head[0], head[1]);
                    self.hash_in_2(tail[0], tail[1]);
                }
```
However, it is much slower (~65000MB/s).
It is worth noting that when compiled without `vaes`, there is also a slow down so I suppose there may be a room for improvement. Could you check it out on your side?